# N-gram Language Models

### Task #1

<u>Download two large text datasets from Kaggle.</u>
* Wine-reviews dataset and All the news dataset 

In [2]:
import time
print('Code started running at {}'.format(time.strftime('%H:%M:%S')))

Code started running at 22:49:46


In [ ]:
import urllib.request

print('Beginning file download...')

url = 'https://storage.googleapis.com/kaggle-data-sets/1974/3493/bundle/archive.zip?GoogleAccessId=web-data@kaggle-161607.iam.gserviceaccount.com&Expires=1583095659&Signature=QHSw6koS1jnqakwJYTHxIvib0ISUJxDwjmx0lEWugNLUBDn1Tc1Pgs2zcR58FjIguEwCr58Z7YEYoOo9z8ycIQEhnmFENNIoEHXR6hX6MP1AiMTCmvDTCGY7xPeTzdobSwuWmr5Cpb7LDPQsolQBoNTftvYItoqsEJEq0cizTUacvW7k4sIG%2BHJDIBmdbYCoG04jq6J8ZgpZJ7zaNP%2ByuyeXJuebdn3IzxZS7KdZ57YC6IIJTzNsdaQJ%2FLxv3mLRglTRyq8GvIekaSkxC6qmJTNydFsPtX2Y6%2FGPcSCR108GmusXo2NZFdJue0grBMStl1D1ZTidJHQQhJA%2FxSfU5A%3D%3D&response-content-disposition=attachment%3B+filename%3Dall-the-news.zip'  

urllib.request.urlretrieve(url, 'all-the-news.zip')

In [ ]:
import zipfile

Project_zip = zipfile.ZipFile('all-the-news.zip')
Project_zip.extract('articles1.csv')
Project_zip.extract('articles2.csv')
Project_zip.extract('articles3.csv')
Project_zip.close()

In [ ]:
import urllib.request

print('Beginning file download...')

url = 'https://storage.googleapis.com/kaggle-data-sets/1442/8172/bundle/archive.zip?GoogleAccessId=web-data@kaggle-161607.iam.gserviceaccount.com&Expires=1583095631&Signature=Saa5KqRhvtI3hXfFDJecE0z7ch%2BloXkWUWvBpzNCE%2Bb0zV5PjNdlohH9RX7qs62RrK6WZwb0Mbg1oY9XVD%2Fbbgs0Lkg5lO%2F5pdb2AOM%2BRy6GQiLE%2F2kcYzFnCus%2BnO68rYJ5z%2B1Sfhmzv5xv%2BiRCcQLnkqrHK9EPfrA%2BUkj2DpubSRsvUoPg8mtra33J4zGtrDzOhslEz0wqRYfsY5rbjKN1Q4nrxs%2BV2ctuGGXfxfbngosOSic1R1JgLhfh1h8qxz3Y7T74FL5r%2FFHgfDqnoYKsF4jfXGkfOcDDBQdaeTLuSLJDnIuaRXrWSYIrCTarRW51aahoHxkGfGJAz%2BJhEQ%3D%3D&response-content-disposition=attachment%3B+filename%3Dwine-reviews.zip'
urllib.request.urlretrieve(url, 'wine-reviews.zip')

In [ ]:
import zipfile

Project_zip = zipfile.ZipFile('wine-reviews.zip')
Project_zip.extract('winemag-data-130k-v2.csv')

Project_zip.close()

In [ ]:
%ls

In [2]:
import nltk

In [3]:
import csv

### Task #2

<u>Process the downloaded <code>csv</code> files in python.</u>

There's a nice csv library already included in python for accessing values in that are stored in a comma separated values (csv) format. Read the <a href="https://docs.python.org/3/library/csv.html">csv library documentation</a>.
What is the delimiter in your csv files? Open each of the two .csv files that you downloaded using this library and be able to read in the data. Note that we really only care about the text column in this assignment.

In [4]:
# Reading the wine-reviews dataset as a csv file and keeping only the text
# column in a list
wine_reviews_list = [] # this list will contain the values of the text column
with open('winemag-data-130k-v2.csv', newline='',encoding="utf8") as wine_reviews_f:
    wine_reviews_csv = csv.reader(wine_reviews_f, delimiter=',', quotechar='"')
    # Skipping the header that contains column names
    next(wine_reviews_csv, None)
    for row in wine_reviews_csv:
        # We choose item with index 2 because the column of index 2 in the
        # dataset is the text column that we want
        wine_reviews_list.append(row[2])
print('There are {} items of wine reviews'.format(len(wine_reviews_list)))

There are 129971 items of wine reviews


In [5]:
# To make later computations faster, we will keep the first 20,000 reviews
wine_reviews_list = wine_reviews_list[:20_000]

In [6]:
# Taking a look at the wine-reviews list: seeing the first two text reviews
print('Examples of wine reviews:')
print('-'*25)
for i,r in enumerate(wine_reviews_list[:2]):
    print(i+1, ') ', r, sep='')
    print()

Examples of wine reviews:
-------------------------
1) Aromas include tropical fruit, broom, brimstone and dried herb. The palate isn't overly expressive, offering unripened apple, citrus and dried sage alongside brisk acidity.

2) This is ripe and fruity, a wine that is smooth while still structured. Firm tannins are filled out with juicy red berry fruits and freshened with acidity. It's  already drinkable, although it will certainly be better from 2016.



In [7]:
# Repeating the same process for the all-the-news datasets
# We will keep only the contents of the text columns in a list
news_list = []

# Because these news datasets contain very long articles, we need to
# increase the limit of the field size
import sys
maxInt = sys.maxsize
while True:
    try:
        csv.field_size_limit(maxInt)
        break
    except OverflowError:
        maxInt = int(maxInt/10)

with open('articles1.csv', newline='',encoding="utf8") as news_f_1:
    news1_csv = csv.reader(news_f_1, delimiter=',', quotechar='"')
    next(news1_csv, None)
    for row in news1_csv:
        news_list.append(row[9])
    
with open('articles2.csv', newline='',encoding="utf8") as news_f_2:
    news2_csv = csv.reader(news_f_2, delimiter=',', quotechar='"')
    next(news2_csv, None)
    for row in news2_csv:
        news_list.append(row[9])
    
with open('articles3.csv', newline='',encoding="utf8") as news_f_3:
    news3_csv = csv.reader(news_f_3, delimiter=',', quotechar='"')
    next(news3_csv, None)
    for row in news3_csv:
        news_list.append(row[9])
        
print('There are {} items of news articles'.format(len(news_list)))

There are 142570 items of news articles


In [8]:
# To make later computations faster, we will keep the first 10000 news articles
news_list = news_list[:10_000]

In [9]:
# Taking a look at the news list: seeing the beginning of two news articles
print('Examples of news articles (only a part of each article is shown):')
print('-'*25)
for i,r in enumerate(news_list[:2]):
    print(i+1, ') ', r[:1000] + '...', sep='')
    print()

Examples of news articles (only a part of each article is shown):
-------------------------
1) WASHINGTON  —   Congressional Republicans have a new fear when it comes to their    health care lawsuit against the Obama administration: They might win. The incoming Trump administration could choose to no longer defend the executive branch against the suit, which challenges the administration’s authority to spend billions of dollars on health insurance subsidies for   and   Americans, handing House Republicans a big victory on    issues. But a sudden loss of the disputed subsidies could conceivably cause the health care program to implode, leaving millions of people without access to health insurance before Republicans have prepared a replacement. That could lead to chaos in the insurance market and spur a political backlash just as Republicans gain full control of the government. To stave off that outcome, Republicans could find themselves in the awkward position of appropriating huge sums

### Task #3

<u>Perform sentence segmentation and word tokenization.</u>

Utilize the nltk module to perform sentence segmentation and word tokenization. But at this point, there are a few decisions that need to be made:

* How we should handle the .csv rows in the previous step? If we ignore row makers, and "lump everything together", how will that effect our language model?
* Do we want to remove punctuation? What is the effect of keeping punctuation in the model?
* Do we want to add sentence boundary markers, such as <samp>&lt;S&gt;</samp> and <samp>&lt;/S&gt;</samp>?</li>
* Should two the words <samp>The</samp> and <samp>the</samp> be treated as the same? What are the effects of doing, or not doing, this?

In [10]:
# Making everything lowercase in wine reviews and news datasets
wine_reviews_list = [x.lower() for x in wine_reviews_list]
news_list = [x.lower() for x in news_list]

In [11]:
# Sentence segmentation and word tokenization
from nltk.tokenize import sent_tokenize, word_tokenize
# For wine reviews
wine_reviews_sentences = [sent_tokenize(x) for x in wine_reviews_list]
wine_reviews_sentences = [item for sublist in wine_reviews_sentences for item in sublist]
# Now the wine_reviews_sentences list is a list of sentences
wine_reviews_words = word_tokenize(' '.join(wine_reviews_sentences))
# For the news
news_sentences = [sent_tokenize(x) for x in news_list]
news_sentences = [item for sublist in news_sentences for item in sublist]
news_words = word_tokenize(' '.join(news_sentences))
# We didn't add sentence boundaries (i.e. <S> and </S>) and kept the punctuation.
# This way, we can now from punctuations when the sentence starts and when it ends
# without the need for sentence boundaries

In [12]:
print('We have {} sentences for wine reviews'.format(len(wine_reviews_sentences)))
print('We have {} words for wine reviews'.format(len(wine_reviews_words)))
print()
print('We have {} sentences for news articles'.format(len(news_sentences)))
print('We have {} words for news articles'.format(len(news_words)))

We have 54273 sentences for wine reviews
We have 940176 words for wine reviews

We have 470477 sentences for news articles
We have 11829689 words for news articles


In [13]:
print('Examples of wine-reviews sentences after segmentation:')
print('-'*20)
for i,w in enumerate(wine_reviews_sentences[:3]):
    print(i+1, ') ', w, sep='')

Examples of wine-reviews sentences after segmentation:
--------------------
1) aromas include tropical fruit, broom, brimstone and dried herb.
2) the palate isn't overly expressive, offering unripened apple, citrus and dried sage alongside brisk acidity.
3) this is ripe and fruity, a wine that is smooth while still structured.


In [14]:
print('Examples of wine-reviews words after tokenization:')
print('-'*20)
for i,w in enumerate(wine_reviews_words[:5]):
    print(i+1, ') ', w, sep='')

Examples of wine-reviews words after tokenization:
--------------------
1) aromas
2) include
3) tropical
4) fruit
5) ,


In [15]:
print('Examples of news sentences after segmentation:')
print('-'*20)
for i,w in enumerate(news_sentences[:3]):
    print(i+1, ') ', w, sep='')

Examples of news sentences after segmentation:
--------------------
1) washington  —   congressional republicans have a new fear when it comes to their    health care lawsuit against the obama administration: they might win.
2) the incoming trump administration could choose to no longer defend the executive branch against the suit, which challenges the administration’s authority to spend billions of dollars on health insurance subsidies for   and   americans, handing house republicans a big victory on    issues.
3) but a sudden loss of the disputed subsidies could conceivably cause the health care program to implode, leaving millions of people without access to health insurance before republicans have prepared a replacement.


In [16]:
print('Examples of news words after tokenization:')
print('-'*20)
for i,w in enumerate(news_words[:5]):
    print(i+1, ') ', w, sep='')

Examples of news words after tokenization:
--------------------
1) washington
2) —
3) congressional
4) republicans
5) have


### Task #4

<u>Calculate N-gram counts and compute probabilities.</u>

Use a python dictionary (or any suitable data structure) to first compute unigram counts. Then try bigram counts. Finally, trigram counts.

How much memory are you using? How fast, or slow, is the code -- how long is this step taking? If it is taking too long, try only using a fraction of your corpus: instead of loading the entire .csv file, try only reading the first 1000 rows of data.

Using those counts, compute the probabilities for the unigrams, bigrams, and trigrams, and store those in a new python dictionary (or some other data structure).

In [17]:
from nltk import bigrams, trigrams
from collections import Counter

# For wine reviews
wine_unigrams = wine_reviews_words.copy()
wine_unigrams_count = Counter(wine_unigrams)
wine_bigrams = bigrams(wine_reviews_words)
wine_bigrams_count = Counter(wine_bigrams)
wine_trigrams = trigrams(wine_reviews_words)
wine_trigrams_count = Counter(wine_trigrams)

# For the news
news_unigrams = news_words.copy()
news_unigrams_count = Counter(news_unigrams)
news_bigrams = bigrams(news_words)
news_bigrams_count = Counter(news_bigrams)
news_trigrams = trigrams(news_words)
news_trigrams_count = Counter(news_trigrams)

### Below are count of Unigram, bigram and trigram of Wine dataset

## 1. Unigram count for wine dataset

In [18]:
wine_unigrams_count

Counter({'aromas': 6188,
         'include': 110,
         'tropical': 575,
         'fruit': 6891,
         ',': 68065,
         'broom': 35,
         'brimstone': 14,
         'and': 53697,
         'dried': 1165,
         'herb': 989,
         '.': 51993,
         'the': 33566,
         'palate': 5803,
         'is': 15060,
         "n't": 571,
         'overly': 82,
         'expressive': 100,
         'offering': 396,
         'unripened': 1,
         'apple': 1841,
         'citrus': 1706,
         'sage': 250,
         'alongside': 873,
         'brisk': 351,
         'acidity': 5401,
         'this': 17427,
         'ripe': 4214,
         'fruity': 1505,
         'a': 27464,
         'wine': 12296,
         'that': 6125,
         'smooth': 1127,
         'while': 1785,
         'still': 857,
         'structured': 794,
         'firm': 1373,
         'tannins': 4691,
         'are': 3868,
         'filled': 46,
         'out': 1159,
         'with': 18669,
         'juicy': 151

## 2. Bigram count for wine dataset

In [19]:
wine_bigrams_count

Counter({('aromas', 'include'): 58,
         ('include', 'tropical'): 3,
         ('tropical', 'fruit'): 283,
         ('fruit', ','): 1230,
         (',', 'broom'): 2,
         ('broom', ','): 18,
         (',', 'brimstone'): 6,
         ('brimstone', 'and'): 5,
         ('and', 'dried'): 271,
         ('dried', 'herb'): 147,
         ('herb', '.'): 127,
         ('.', 'the'): 7834,
         ('the', 'palate'): 3581,
         ('palate', 'is'): 1005,
         ('is', "n't"): 97,
         ("n't", 'overly'): 8,
         ('overly', 'expressive'): 2,
         ('expressive', ','): 31,
         (',', 'offering'): 244,
         ('offering', 'unripened'): 1,
         ('unripened', 'apple'): 1,
         ('apple', ','): 428,
         (',', 'citrus'): 242,
         ('citrus', 'and'): 439,
         ('dried', 'sage'): 31,
         ('sage', 'alongside'): 19,
         ('alongside', 'brisk'): 18,
         ('brisk', 'acidity'): 118,
         ('acidity', '.'): 1741,
         ('.', 'this'): 5704,
         

## 3. Trigram count for wine dataset

In [20]:
wine_trigrams_count

Counter({('aromas', 'include', 'tropical'): 1,
         ('include', 'tropical', 'fruit'): 2,
         ('tropical', 'fruit', ','): 72,
         ('fruit', ',', 'broom'): 1,
         (',', 'broom', ','): 2,
         ('broom', ',', 'brimstone'): 2,
         (',', 'brimstone', 'and'): 4,
         ('brimstone', 'and', 'dried'): 1,
         ('and', 'dried', 'herb'): 41,
         ('dried', 'herb', '.'): 19,
         ('herb', '.', 'the'): 54,
         ('.', 'the', 'palate'): 1454,
         ('the', 'palate', 'is'): 867,
         ('palate', 'is', "n't"): 10,
         ('is', "n't", 'overly'): 4,
         ("n't", 'overly', 'expressive'): 1,
         ('overly', 'expressive', ','): 1,
         ('expressive', ',', 'offering'): 5,
         (',', 'offering', 'unripened'): 1,
         ('offering', 'unripened', 'apple'): 1,
         ('unripened', 'apple', ','): 1,
         ('apple', ',', 'citrus'): 30,
         (',', 'citrus', 'and'): 129,
         ('citrus', 'and', 'dried'): 1,
         ('and', 'dried', 

### Below are counts of unigram, bigram and trigram for all the news dataset

## 1. Unigram count for all the news dataset

In [21]:
news_unigrams_count

Counter({'washington': 4851,
         '—': 42136,
         'congressional': 1007,
         'republicans': 3707,
         'have': 44759,
         'a': 262228,
         'new': 21971,
         'fear': 1213,
         'when': 21396,
         'it': 79042,
         'comes': 1770,
         'to': 269860,
         'their': 27363,
         'health': 4562,
         'care': 3625,
         'lawsuit': 865,
         'against': 8129,
         'the': 594538,
         'obama': 7955,
         'administration': 5588,
         ':': 22622,
         'they': 39469,
         'might': 5049,
         'win': 2056,
         '.': 455147,
         'incoming': 294,
         'trump': 37638,
         'could': 13727,
         'choose': 560,
         'no': 14436,
         'longer': 1573,
         'defend': 604,
         'executive': 4015,
         'branch': 384,
         'suit': 630,
         ',': 640101,
         'which': 21861,
         'challenges': 734,
         '’': 169755,
         's': 113572,
         'authority':

## 2. Bigram count for all the news dataset

In [22]:
news_bigrams_count

Counter({('washington', '—'): 913,
         ('—', 'congressional'): 5,
         ('congressional', 'republicans'): 102,
         ('republicans', 'have'): 242,
         ('have', 'a'): 3141,
         ('a', 'new'): 3729,
         ('new', 'fear'): 1,
         ('fear', 'when'): 2,
         ('when', 'it'): 1289,
         ('it', 'comes'): 457,
         ('comes', 'to'): 441,
         ('to', 'their'): 1181,
         ('their', 'health'): 78,
         ('health', 'care'): 1302,
         ('care', 'lawsuit'): 1,
         ('lawsuit', 'against'): 108,
         ('against', 'the'): 2118,
         ('the', 'obama'): 1440,
         ('obama', 'administration'): 1390,
         ('administration', ':'): 15,
         (':', 'they'): 227,
         ('they', 'might'): 219,
         ('might', 'win'): 11,
         ('win', '.'): 255,
         ('.', 'the'): 39891,
         ('the', 'incoming'): 187,
         ('incoming', 'trump'): 35,
         ('trump', 'administration'): 1134,
         ('administration', 'could'): 56,
 

## 3. Trigram count for all the news dataset

In [23]:
news_trigrams_count

Counter({('washington', '—', 'congressional'): 5,
         ('—', 'congressional', 'republicans'): 4,
         ('congressional', 'republicans', 'have'): 11,
         ('republicans', 'have', 'a'): 5,
         ('have', 'a', 'new'): 19,
         ('a', 'new', 'fear'): 1,
         ('new', 'fear', 'when'): 1,
         ('fear', 'when', 'it'): 1,
         ('when', 'it', 'comes'): 345,
         ('it', 'comes', 'to'): 336,
         ('comes', 'to', 'their'): 5,
         ('to', 'their', 'health'): 1,
         ('their', 'health', 'care'): 19,
         ('health', 'care', 'lawsuit'): 1,
         ('care', 'lawsuit', 'against'): 1,
         ('lawsuit', 'against', 'the'): 44,
         ('against', 'the', 'obama'): 11,
         ('the', 'obama', 'administration'): 1226,
         ('obama', 'administration', ':'): 2,
         ('administration', ':', 'they'): 1,
         (':', 'they', 'might'): 1,
         ('they', 'might', 'win'): 1,
         ('might', 'win', '.'): 3,
         ('win', '.', 'the'): 21,
       

In [24]:
# To compute the probabilities, we will use a nltk functions
# named FreqDist, ConditionalFreqDist, and ConditionalProbDist
from nltk import FreqDist, ConditionalFreqDist, ConditionalProbDist, ELEProbDist
# For wine-reviews dataset

# First for unigrams
wine_fdist_uni = FreqDist(wine_unigrams)
wine_pdist_uni = {}
wine_uni_sum = sum(wine_fdist_uni.values())
for k,v in wine_fdist_uni.items():
    wine_pdist_uni[k] = v / wine_uni_sum
# Then for bigrams
wine_cfdist_bi = ConditionalFreqDist(bigrams(wine_reviews_words))
wine_cpdist_bi = ConditionalProbDist(wine_cfdist_bi, ELEProbDist)
# Then for trigrams
wine_condition_pairs = (((w0, w1), w2) for w0, w1, w2 in trigrams(wine_reviews_words))
wine_cfdist_tri = ConditionalFreqDist(wine_condition_pairs)
wine_cpdist_tri = ConditionalProbDist(wine_cfdist_tri, ELEProbDist)

# For news dataset
# First for unigrams
news_fdist_uni = FreqDist(news_unigrams)
news_pdist_uni = {}
news_uni_sum = sum(news_fdist_uni.values())
for k,v in news_fdist_uni.items():
    news_pdist_uni[k] = v / news_uni_sum
# Then for bigrams
news_cfdist_bi = ConditionalFreqDist(bigrams(news_words))
news_cpdist_bi = ConditionalProbDist(news_cfdist_bi, ELEProbDist)
# Then for trigrams
news_condition_pairs = (((w0, w1), w2) for w0, w1, w2 in trigrams(news_words))
news_cfdist_tri = ConditionalFreqDist(news_condition_pairs)
news_cpdist_tri = ConditionalProbDist(news_cfdist_tri, ELEProbDist)

#### As the data is large, everytime we tried to print output it gave error of cannot handle large data. So we just printed probabilities of Unigrams for both dataset

In [25]:
wine_pdist_uni

{'aromas': 0.006581746396419394,
 'include': 0.00011699937033066149,
 'tropical': 0.0006115876176375487,
 'fruit': 0.0073294787358962575,
 ',': 0.07239601946869523,
 'broom': 3.722707237793775e-05,
 'brimstone': 1.4890828951175099e-05,
 'and': 0.05711377444223209,
 'dried': 0.001239129694865642,
 'herb': 0.0010519307023365838,
 '.': 0.055301347832746206,
 'the': 0.03570182604108167,
 'palate': 0.006172248600262079,
 'is': 0.016018277428906928,
 "n't": 0.0006073330950800702,
 'overly': 8.721771242831129e-05,
 'expressive': 0.000106363063936965,
 'offering': 0.0004211977331903814,
 'unripened': 1.06363063936965e-06,
 'apple': 0.0019581440070795255,
 'citrus': 0.0018145538707646227,
 'sage': 0.0002659076598424125,
 'alongside': 0.0009285495481697044,
 'brisk': 0.00037333435441874713,
 'acidity': 0.005744669083235479,
 'this': 0.01853589115229489,
 'ripe': 0.004482139514303705,
 'fruity': 0.0016007641122513232,
 'a': 0.029211551879648065,
 'wine': 0.013078402341689216,
 'that': 0.006514737

In [26]:
news_pdist_uni

{'washington': 0.0004100699519657702,
 '—': 0.0035618856928529567,
 'congressional': 8.512480759215226e-05,
 'republicans': 0.00031336411295343437,
 'have': 0.003783615951357639,
 'a': 0.022166939468991956,
 'new': 0.0018572762141084182,
 'fear': 0.00010253862126045748,
 'when': 0.0018086696953740712,
 'it': 0.006681663397913504,
 'comes': 0.00014962354462572938,
 'to': 0.02281209590547985,
 'their': 0.0023130785602225048,
 'health': 0.00038563989298450704,
 'care': 0.00030643240071653616,
 'lawsuit': 7.312111079167001e-05,
 'against': 0.0006871693752895787,
 'the': 0.0502581259744022,
 'obama': 0.0006724606200551849,
 'administration': 0.00047237082902179425,
 ':': 0.0019123072466233051,
 'they': 0.003336435979001646,
 'might': 0.00042680750102559755,
 'win': 0.00017380000437881333,
 '.': 0.0384749759693598,
 'incoming': 2.485272436156183e-05,
 'trump': 0.0031816559167362726,
 'could': 0.0011603855350719701,
 'choose': 4.7338522593451104e-05,
 'no': 0.0012203194859983218,
 'longer': 0

### Task #5

<u>Compare the statistics of the corpora.</u>
                        
Use the results of those calculations that you just made the poor computer painstakingly compute. What are the differences in the most common unigrams between the two language models? Are there interesting differences between the bigram models or trigram models?

Be able to sort the n-grams to output the top k with the highest count or probability.

In [27]:
# Most common unigrams
# For wine reviews
print('Most common unigrams')
print('-'*30)
print('Wine reviews:')
for k,v in wine_unigrams_count.most_common(15):
    print(k.ljust(10), v)
print('='*20)
# For the news
print('News articles:')
for k,v in news_unigrams_count.most_common(15):
    print(k.ljust(10), v)

Most common unigrams
------------------------------
Wine reviews:
,          68065
and        53697
.          51993
the        33566
a          27464
of         26396
with       18669
this       17427
is         15060
it         13117
wine       12296
in         9693
flavors    9655
to         8502
's         7837
News articles:
,          640101
the        594538
.          455147
to         269860
of         263508
a          262228
and        239937
in         213503
’          169755
that       139628
”          137357
“          137350
s          113572
for        95168
on         89836


In [28]:
# Most common bigrams
# For wine reviews
print('Most common bigrams')
print('-'*30)
print('Wine reviews:')
for k,v in wine_bigrams_count.most_common(15):
    print(' '.join(k).ljust(10), v)
print('='*20)
# For the news
print('News articles:')
for k,v in news_bigrams_count.most_common(15):
    print(' '.join(k).ljust(10), v)

Most common bigrams
------------------------------
Wine reviews:
. the      7834
. it       6691
, with     6582
. this     5704
, this     5288
on the     5044
it 's      4566
this is    3821
the palate 3581
with a     3433
and a      3353
is a       3188
. a        3140
. drink    3106
finish .   2935
News articles:
’ s        113252
. “        66092
of the     61696
, ”        55516
in the     49235
. ”        48619
, the      42623
. the      39891
, and      38875
, a        27228
to the     25342
said .     22210
’ t        21959
mr. trump  21059
on the     20764


In [29]:
# Most common trigrams
# For wine reviews
print('Most common trigrams')
print('-'*30)
print('Wine reviews:')
for k,v in wine_trigrams_count.most_common(15):
    print(' '.join(k).ljust(18), v)
print('='*20)
# For the news
print('News articles:')
for k,v in news_trigrams_count.most_common(15):
    print(' '.join(k).ljust(18), v)

Most common trigrams
------------------------------
Wine reviews:
. it 's            3054
this is a          1960
. this is          1944
, this is          1480
. the palate       1454
, with a           1420
the finish .       1389
on the finish      1307
on the palate      1301
on the nose        1153
. the wine         1020
. drink now        986
the palate ,       948
drink now .        907
. it has           905
News articles:
it ’ s             12552
said . “           11134
, ” he             10550
. “ i              9982
, ” said           9081
trump ’ s          8730
the united states  8603
” he said          8458
, ” mr.            7758
don ’ t            7230
he said .          6656
mr. trump ’        6112
. “ it             5788
. ” the            5490
. “ we             5477


### Task #6

<u>Generate random sentences from the N-grams models for both datasets.</u>
                        
We briefly talked about this idea in class. It's also introduced at a high-level in J&M 4.3. How can a random number in the range [0,1] probabilistically generate a word using your model?

In [30]:
# For wine reviews
# First we choose a random word to start the sentence
import random
print('Examples of randomly generated wine-reviews sentences (based on a probability distribution): ')
print('-'*60)
for i in range(3):
    print(i+1, ') ', sep='', end='')
    first_word = None
    for k,v in wine_pdist_uni.items():
        if round(v, 2) == round(random.random(), 2):
            first_word = k
            break
    print(first_word[0].upper() + first_word[1:], end=' ')
    # Now we generate next words: each word based on the word before it.
    # The function generate() returns a word based on its probability.
    # So if the word "good" comes after the word "really" with a probability
    # of 0.5, then generate() will return "good" after "really" 50% of the time
    current_word = None
    prev_word = first_word
    while current_word != '.':
        current_word = wine_cpdist_bi[prev_word].generate()
        prev_word = current_word
        print(current_word, end=' ')
    print()

Examples of randomly generated wine-reviews sentences (based on a probability distribution): 
------------------------------------------------------------
1) Off on complexity , has solid with perhaps still kicking in this crisp and vanilla and has all framed riesling comes from young , foxy ” —light and leather aromas include spice and black currant , zin . 
2) Direct line offers flavors of apple juice . 
3) Off-dry , prune and a true flavors are jumbled aromas dress up the vintage blends , almost cleansing burst through to the vines , nebbiolo-based wines , this aged in . 


In [31]:
# For news, we repeat the same process
print('Examples of randomly generated news sentences (based on a probability distribution): ')
print('-'*60)
for i in range(3):
    print(i+1, ') ', sep='', end='')
    first_word = None
    for k,v in news_pdist_uni.items():
        if round(v, 2) == round(random.random(), 2):
            first_word = k
            break
    print(first_word[0].upper() + first_word[1:], end=' ')
    current_word = None
    prev_word = first_word
    while current_word != '.':
        current_word = news_cpdist_bi[prev_word].generate()
        prev_word = current_word
        print(current_word, end=' ')
    print()

Examples of randomly generated news sentences (based on a probability distribution): 
------------------------------------------------------------
1) Dim his cousins and instead , ” with material to provide $ 3 a. m. if we needed . 
2) Executive vice president trump named dan gray hairs of argument . 
3) Well . 


In [1]:
#to see how much time the whole notebook took
print('Code finished running at {}'.format(time.strftime('%H:%M:%S')))

NameError: name 'time' is not defined

### Report

### The Report

In this project, we built two language models out of two datasets. For each model, we 

* processed the data with segmentation and tokenization, 
* extracted the unigrams, bigrams, and trigrams, 
* found the probability distributions for the n-grams, 
* compared the outputs of the two models
* generated random sentences using the probability distributions of each model

In the following, we explain different aspects of the process with notes and comments.

#### Datasets Used

In this work, two datasets from Kaggle were used and compared:

1. Wine Reviews dataset
2. All the News dataset

#### Amount of Data Used

At first, the whole datasets were included in the analysis. However, it turned out that this would consume a lot of time so the first 20,000 items of the wine-reviews dataset and the first 10,000 articles of the news-articles dataset were kept only.

#### Size of the Datasets

* The wine-reviews dataset contains originally 129,971 reviews. Because this big amount of data makes operations take a lot of time, we used the first 20,000 reviews
* The news-articles dataset contains originally 142,570 reviews. Also, because this big amount of data makes operations take a lot of time, we used the first 10,000 articles

Regarding the number of sentences that we got after segmentation for each dataset:

* We got 54,273 sentences for the wine-reviews dataset
* We got 470,477 sentences for the news-articles dataset

Regarding the number of words that we got after tokenization for each dataset:

* We got 940,176 words for the wine-reviews dataset
* We got 11,829,689 words for the news-articles dataset

#### Processing the Data

For handling the wine-reviews items, we used NLTK's `sent_tokenize()` funtion to sentence-segment each review and then, we put all the sentences in one list. We did the same for the news articles as shown in the code section above.

After that, we word-tokenized the string that results from concatenating the segmented sentences from the previous step.

In the tokenization and segmentation processes, we kept the punctuation marks and didn't add sentence boundaries. This way, we can know from punctuations when the sentence starts and when it ends without the need for sentence boundaries (i.e. the sentence starts after a period and ends before a period too).

#### Time Needed to Run this Program

When we tried to use the whole datasets, the tokenization and segmentation processes in Task3 took a lot of time. So, we kept a part of each dataset as explained above. With this, the code in this notebook took around 9 minutes to finish running.

#### Top Unigrams, Bigrams, and Trigrams for each Model

We showed the top unigrams, bigrams, and trigrams for each model above. We will list them again here:

**Most common unigrams with the number of occurrences (Wine-reviews model):**

|Unigram|No. of occurrences|
|---|---|
|,|          68065|
|and|        53697|
|.|          51993|
|the|        33566|
|a|          27464|
|of|         26396|
|with|       18669|
|this|       17427|
|is|         15060|
|it|         13117|
|wine|       12296|
|in|         9693|
|flavors|    9655|
|to|         8502|
|'s|         7837|

**Most common unigrams with the number of occurrences (News-articles model):**

|Unigram|No. of occurrences|
|---|---|
|,|          640101|
|the |       594538|
|.|          455147|
|to|         269860|
|of|         263508|
|a|          262228|
|and|        239937|
|in|         213503|
|’|          169755|
|that|       139628|
|”|          137357|
|“|          137350|
|s|          113572|
|for|        95168|
|on|         89836|

We can see the most of the most common unigrams for both language models are very common English words such as "the", "and", "for", etc. We can remove these by excluding stop-words from the text.

We also see that the words "wine" and "flavors" are among the most common unigrams of the wine-reviews model.

**Most common bigrams with the number of occurrences (Wine-reviews model):**

|Bigram|No. of occurrences|
|---|---|
|. the      |7834|
|. it       |6691|
|, with     |6582|
|. this     |5704|
|, this     |5288|
|on the     |5044|
|it 's      |4566|
|this is    |3821|
|the palate |3581|
|with a     |3433|
|and a      |3353|
|is a       |3188|
|. a        |3140|
|. drink    |3106|
|finish .   |2935|

**Most common bigrams with the number of occurrences (News-articles model):**

|Bigram|No. of occurrences|
|---|---|
|’ s        |113252|
|. “        |66092|
|of the     |61696|
|, ”        |55516|
|in the     |49235|
|. ”        |48619|
|, the      |42623|
|. the      |39891|
|, and      |38875|
|, a        |27228|
|to the     |25342|
|said .     |22210|
|’ t        |21959|
|mr. trump  |21059|
|on the     |20764|

We notice that "the palate", ". drink", and "finish ." are among the most common bigrams of the wine-reviews language model. On the other hand, we notice that "said ." and "mr. trump" are among the most common bigrams of the news model.

**Most common trigrams with the number of occurrences (Wine-reviews model):**

|Trigram|No. of occurrences|
|---|---|
|. it 's            |3054|
|this is a          |1960|
|. this is          |1944|
|, this is          |1480|
|. the palate       |1454|
|, with a           |1420|
|the finish .       |1389|
|on the finish      |1307|
|on the palate      |1301|
|on the nose        |1153|
|. the wine         |1020|
|. drink now        |986|
|the palate ,       |948|
|drink now .        |907|
|. it has           |905|

**Most common trigrams with the number of occurrences (News-articles model):**

|Trigram|No. of occurrences|
|---|---|
|it ’ s             |12552|
|said . “           |11134|
|, ” he             |10550|
|. “ i              |9982|
|, ” said           |9081|
|trump ’ s          |8730|
|the united states  |8603|
|” he said          |8458|
|, ” mr.            |7758|
|don ’ t          |  |7230|
|he said .          |6656|
|mr. trump ’        |6112|
|. “ it             |5788|
|. ” the            |5490|
|. “ we             |5477|

For trigrams, the differences between the two models become clearer. For the wine model, we see ". the palate", "the finish .", "on the finish", "on the palate", "on the nose", ". the wine", and "drink now ." among the most common trigrams. On the other hand, we see "said . “", ", ” said", "trump ’ s", "the united states", "” he said", ", ” mr.", "he said .", and "mr. trump ’" among the most common trigrams.

#### Examples of Randomly Generated Sentences for each Model

We used NLTK's utilities to randomly generate sentences using probabiliy distributions as shown in the code of Task6. Here, we will list 3 example sentences for each language model.

**Wine-reviews random sentences:**
1. Of spice and medium bodied but does amazing thing ; it 's 2009 and honey , fully drinkable wine is prominent , freshly shaved mineral flavors that provides necessary . 
2. Layered and concentrated . 
3. Tropical fruit , and a bit sweet and rich in tightly packed with foresty berry , winside vineyard expression of clove accents will develop well through 2020. plum , star .

**News-articles random sentences:**
1. Handle classified information to influence efforts to pick to competitive and high school of guantánamo , for decisive ” which five days of preparing for secretary of sly on busy cooking , for subprime mortgages , ” which was one of financial backers that republicans since 2000 , money ebbs . 
2. Preserving and our goals . 
3. — that the top of the study in court in cash transfers announced only in 2010 email that both attacks in shanghai noon on earth institute of millions of islands with jeff rosen college republicans talk to the first three days during a tweet referred to reverse that vouchers from some aspect of the huge victory for a marathon sessions brings foreign secretary of california coast would help jog . 

We notice that the sentences contain some understandable phrases and we expect that with more advanced techniques, better results can be achieved.